rpc,security, cli: Add cert-distinguished-name root/node cli flag

souravcrl opened the pull request:
Previous in sequence: https://github.com/cockroachdb/cockroach/pull/119958
informs #110616 
fixes #118750
fixes CRDB-35884
Epic CRDB-34126

We will be adding 2 new cli flags `root-cert-distinguished-name` and
`node-cert-distinguished-name` to provide option to have subject DN for root and
node user during server startup. This will enforce the provided certificate by
client to exactly match the value set by the above flags both for sql client and
RPC authentication. This is needed because subject role option cannot be set for
root and node users.

Post this the plan is to add a cluster setting
`server.client_cert.subject_required` which will mandate that any auth which
happens should verify certSubject with rootSubject in case of root user, with
nodeSubject in case of node user, with roleSubject otherwise.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120786)
<!-- Reviewable:end -->

bdarnell reviewed (commented):
> Reviewed 7 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @souravcrl)

___
*[`pkg/cli/cliflags/flags.go` line 880 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu6jQi-ZzWjLsPEKm6:-Ntu6jQi-ZzWjLsPEKm7:b-vehqkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/cli/cliflags/flags.go#L880)):*
> ```Go
> 		Description: `
> A string of comma separated list of distinguished-name
> <attribute>:<attribute-value> mappings in accordance with RFC2253 for the root
> ```

Should be `=` instead of `:`, right?
___
*[`pkg/cli/cliflags/flags.go` line 882 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu6vQc2vh22wc-nLPL:-Ntu6vQc2vh22wc-nLPM:bt7i4de) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/cli/cliflags/flags.go#L882)):*
> ```Go
> <attribute>:<attribute-value> mappings in accordance with RFC2253 for the root
> user. This strictly needs to match the DN subject in the client certificate
> provided for root user in-case this value is set here.
> ```

No hyphen in "in case". I'd say "provided for the root user if this flag is set".
___
*[`pkg/cli/interactive_tests/test_cert_advisory_validation.tcl` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu7_8h6Z0FV9AzSZJ2:-Ntu7_8h6Z0FV9AzSZJ3:b-vbbw41) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/cli/interactive_tests/test_cert_advisory_validation.tcl#L55)):*
> ```Tcl
> eexpect $prompt
> 
> send "$argv start-single-node --store=$db_dir --certs-dir=$certs_dir --root-cert-distinguished-name=$root_dn --node-cert-distinguished-name=$node_dn --advertise-addr=localhost\r"
> ```

We test that the server can start, but we don't do anything with it. I don't think this test is really doing anything interesting if we don't try to actually log in. (The `create-client root` above is just creating a plain cert that is ever used). 

This test is about the "advisory validation", that is, the warnings that we print when the certs aren't quite right. I would make a new tcl file (copied from the ocsp test to start) for distinguished name validation. 
___
*[`pkg/rpc/auth.go` line 319 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu895l4gOS7fB8Iwjp:-Ntu895l4gOS7fB8Iwjq:b-oot6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/rpc/auth.go#L319)):*
> ```Go
> 	// example, there is no reason to permit the root user to send raw
> 	// Raft RPCs.
> 	if !security.CheckCertDNMatchesRootDNorNodeDN(clientCert) {
> ```

This doesn't look right - it treats "distinguished name flag not set" the same as "distinguished name flag set but cert doesn't match". We only want to fall back to GetCertificateUserScope if the flag is unset. If the flag is set and doesn't match we fail immediately. 
___
*[`pkg/security/auth.go` line 290 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu90ka5mZ3AznK08T3:-Ntu90ka5mZ3AznK08T4:bhf6n5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/security/auth.go#L290)):*
> ```Go
> 		}
> 
> 		if roleSubject == nil {
> ```

So if someone were to insert a subject role option for root, that would take precedence over the one on the command-line flag. I think I'd prefer to have the flag take precedence: the ability to change command-line flags when restarting the server is one way to recover from locking yourself out. So instead of `if roleSubject == nil` we should use `getRootOrNodeSubject` if `systemIdentity` is `root` or `node`. 
___
*[`pkg/rpc/auth_test.go` line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu9ZEXELTvGD1joYRD:-Ntu9ZEXELTvGD1joYRE:bmy5xic) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/rpc/auth_test.go#L253)):*
> ```Go
> 				require.Regexp(t, tc.expErr, err)
> 			}
> 			// post test execution unset root and node DN set
> ```

Use `defer` to ensure that these get run even if the test panics (and then we can put the `defer` right next to the `Set*Subject` call)
___
*[`pkg/security/auth_test.go` line 341 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-NtuAH-96mLBbvGmnY77:-NtuAH-96mLBbvGmnY78:bo0ntzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/security/auth_test.go#L341)):*
> ```Go
> 	subsetSubjectDNString := "O=Cockroach,OU=Order Processing Team,CN=foo"
> 	fieldMismatchOnlyOnCommonNameString := "O=Cockroach,OU=Order Processing Team,UID=b8b40653-7f74-4f14-8a61-59f7f3b18184,CN=bar"
> 	rootDNString := "O=Cockroach,OU=Order Processing Team,UID=b8b40653-7f74-4f14-8a61-59f7f3b18184,CN=root"
> ```

These DN strings with CN=root (here and in the tcl test) are not good test cases because they'd be accepted even if the DN flags were ignored. We need test cases that show that mismatches in the other DN fields will cause the cert to be rejected. 


<!-- Sent from Reviewable.io -->

souravcrl committed:
rpc,security, cli: Add cert-distinguished-name root/node cli flag

Previous in sequence: https://github.com/cockroachdb/cockroach/pull/119958
informs #110616, #118750
fixes CRDB-35884
Epic CRDB-34126

We will be adding 2 new cli flags `root-cert-distinguished-name` and
`node-cert-distinguished-name` to provide option to have subject DN for root and
node user during server startup. This will enforce that certificate provided by
client exactly matches the value set by the above flags both for sql client and
RPC authentication. This is needed because subject role option cannot be set for
root and node users.

Post this the plan is to add a cluster setting
`server.client_cert.subject_required` which will mandate that any auth which
happens should verify certSubject with rootSubject in case of root user, with
nodeSubject in case of node user, with roleSubject otherwise.

Release note: None

souravcrl reviewed (commented):
PTAL. I have added a new DN validation flag and fixed other comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-:-Nu499Gj3dZXKZ1YEuJH:bxump4c)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @souravcrl)

___
*[`pkg/cli/cliflags/flags.go` line 880 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu6jQi-ZzWjLsPEKm6:-Nu3PbvR55GDvRdMXCci:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/cli/cliflags/flags.go#L880)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Should be `=` instead of `:`, right?
</blockquote></details>

done
___
*[`pkg/cli/cliflags/flags.go` line 882 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu6vQc2vh22wc-nLPL:-Nu3PcqtEPC6UmKzoYkg:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/cli/cliflags/flags.go#L882)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

No hyphen in "in case". I'd say "provided for the root user if this flag is set".
</blockquote></details>

done
___
*[`pkg/rpc/auth.go` line 319 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu895l4gOS7fB8Iwjp:-Nu3S34R5iNwVHedwYX1:b-4hl0x8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/rpc/auth.go#L319)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This doesn't look right - it treats "distinguished name flag not set" the same as "distinguished name flag set but cert doesn't match". We only want to fall back to GetCertificateUserScope if the flag is unset. If the flag is set and doesn't match we fail immediately.
</blockquote></details>

updated `GetCertificateUserScope` to return rootOrNodeDNSet, certDNMatchesRootorNodeDN so both could be evaluated.
___
*[`pkg/security/auth.go` line 290 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu90ka5mZ3AznK08T3:-Nu3VH3r87sk8ow79lOa:b-bcz5uv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/security/auth.go#L290)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

So if someone were to insert a subject role option for root, that would take precedence over the one on the command-line flag. I think I'd prefer to have the flag take precedence: the ability to change command-line flags when restarting the server is one way to recover from locking yourself out. So instead of `if roleSubject == nil` we should use `getRootOrNodeSubject` if `systemIdentity` is `root` or `node`.
</blockquote></details>

updated `getRootOrNodeSubject` to `evaluateRootOrNodeDN` which is doing this handling.
___
*[`pkg/rpc/auth_test.go` line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu9ZEXELTvGD1joYRD:-Nu3Wg0b8jhpkuxyXRuB:b-3px3mx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/rpc/auth_test.go#L253)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Use `defer` to ensure that these get run even if the test panics (and then we can put the `defer` right next to the `Set*Subject` call)
</blockquote></details>

updated to defer
___
*[`pkg/security/auth_test.go` line 341 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-NtuAH-96mLBbvGmnY77:-Nu3bWYLBy70od0lLTUf:bp1tsnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/security/auth_test.go#L341)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

These DN strings with CN=root (here and in the tcl test) are not good test cases because they'd be accepted even if the DN flags were ignored. We need test cases that show that mismatches in the other DN fields will cause the cert to be rejected.
</blockquote></details>

I have added the test where CN can be different. I realized `ValidateUserScope` had an error where even if certDN matched systemIdDN, it still failed auth since CN did not match. I have updated the implementation to bypass this.
___
*[`pkg/cli/interactive_tests/test_cert_advisory_validation.tcl` line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Ntu7_8h6Z0FV9AzSZJ2:-Nu3PeBjBgaINE81jEkM:b-ot9cru) ([raw file](https://github.com/cockroachdb/cockroach/blob/d196de436149b1ab74dc562c58fc5d3d80096f55/pkg/cli/interactive_tests/test_cert_advisory_validation.tcl#L55)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We test that the server can start, but we don't do anything with it. I don't think this test is really doing anything interesting if we don't try to actually log in. (The `create-client root` above is just creating a plain cert that is ever used). 

This test is about the "advisory validation", that is, the warnings that we print when the certs aren't quite right. I would make a new tcl file (copied from the ocsp test to start) for distinguished name validation.
</blockquote></details>

added a new test file `test_distinguished_name_validation`. removed the tests from here since validation is already handled in the mentioned file test_distinguished_name_validation.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
Only the license key comment is blocking; If you want to get this in before the branch cut you can leave the rest of the comments for a follow-up PR. 

> Reviewed 19 of 19 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-:-Nu4nOtS6lTpeVBsZa_l:b-90gu4t)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @souravcrl)

___
*[`pkg/cli/interactive_tests/test_distinguished_name_validation.tcl` line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4llrE7YLNMjv8HYdV:-Nu4llrFB160HgLUxhAl:b-8kymz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/cli/interactive_tests/test_distinguished_name_validation.tcl#L17)):*
> ```Tcl
> 
> 
> send "mkdir -p $certs_dir\r"
> ```

FYI `system` is preferred over `send` most of the time because you don't need the `\r` at the end and you don't need to follow every command with `eexpect $prompt`. You only need to use `send` when you're doing something interactive and the command needs some input before it will complete. 
___
*[`pkg/cli/interactive_tests/test_distinguished_name_validation.tcl` line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4lQTo8xIFWBNgYSBk:-Nu4lQTo8xIFWBNgYSBl:b-g0hpmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/cli/interactive_tests/test_distinguished_name_validation.tcl#L27)):*
> ```Tcl
> # Copy openssl CA cnf for generating custom client certs.
> set ca_cnf_file [file join [file dirname $argv0] "ocsp_ca.cnf"]
> send "cat $ca_cnf_file > $custom_ca_dir/ca.cnf\r"
> ```

Why not `cp`? or the `file copy` command as used in the ocsp test?
___
*[`pkg/cli/interactive_tests/test_distinguished_name_validation.tcl` line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4mB5s5xpb3JULQtlr:-Nu4mB5s5xpb3JULQtls:b4n1epe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/cli/interactive_tests/test_distinguished_name_validation.tcl#L43)):*
> ```Tcl
>     system "$argv start-single-node --host=localhost --socket-dir=. --certs-dir=$certs_dir --store=$::db_dir --pid-file=server_pid --background $extra >>expect-cmd.log 2>&1;
>             $argv sql --certs-dir=$certs_dir -e 'select 1';
>             $argv sql --certs-dir=$certs_dir -e 'SET CLUSTER SETTING enterprise.license = \"crl-0-EJL04ukFGAEiI0NvY2tyb2FjaCBMYWJzIC0gUHJvZHVjdGlvbiBUZXN0aW5n\"';
> ```

We don't check license keys into the source. If you need a license key for this test, get it from `$COCKROACH_DEV_LICENSE`, which is set in CI and you can set it on your laptop. But I don't think you need a license for this test?
___
*[`pkg/security/auth.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4hI8BDLXImZrRf8WI:-Nu4hI8BDLXImZrRf8WJ:b-rswvfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/security/auth.go#L101)):*
> ```Go
> // original subject option if it is set for the role. Root and Node roles cannot
> // have subject role option set for them.
> func evaluateRootOrNodeDN(
> ```

I like this design but I'd rename the method to something like `applyRootOrNodeDNFlag` to emphasize that it's about combining the command line flags with a maybe-existing role option (it's good to cover that case even though I think it's not required at this point).
___
*[`pkg/security/auth.go` line 142 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4hl4k7jPGNDnBfqPN:-Nu4hl4k7jPGNDnBfqPO:bdt4q0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/security/auth.go#L142)):*
> ```Go
> }
> 
> func SetNodeSubject(nodeDNString string) error {
> ```

Nit: Move the node subject functions up to be next to the root ones. 


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
PTAL. I have updated it but waiting for CI to pass.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-:-Nu4tlRZ-pIUKcXf3ckV:br0utp0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell)

___
*[`pkg/cli/interactive_tests/test_distinguished_name_validation.tcl` line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4llrE7YLNMjv8HYdV:-Nu4q4Vz4JaXLbjZbDrR:b3darh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/cli/interactive_tests/test_distinguished_name_validation.tcl#L17)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

FYI `system` is preferred over `send` most of the time because you don't need the `\r` at the end and you don't need to follow every command with `eexpect $prompt`. You only need to use `send` when you're doing something interactive and the command needs some input before it will complete.
</blockquote></details>

There was a case when the directory was not created and test proceeded with creating a ca certificate. So have added send and was waiting for `prompt` here. 
___
*[`pkg/cli/interactive_tests/test_distinguished_name_validation.tcl` line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4lQTo8xIFWBNgYSBk:-Nu4smbCBJ26OCSHWmg9:b-512vsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/cli/interactive_tests/test_distinguished_name_validation.tcl#L27)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why not `cp`? or the `file copy` command as used in the ocsp test?
</blockquote></details>

updated to cp.
___
*[`pkg/cli/interactive_tests/test_distinguished_name_validation.tcl` line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4mB5s5xpb3JULQtlr:-Nu4sop4EFpNq1ChC3QY:bvnrpcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/cli/interactive_tests/test_distinguished_name_validation.tcl#L43)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We don't check license keys into the source. If you need a license key for this test, get it from `$COCKROACH_DEV_LICENSE`, which is set in CI and you can set it on your laptop. But I don't think you need a license for this test?
</blockquote></details>

updated
___
*[`pkg/security/auth.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4hI8BDLXImZrRf8WI:-Nu4sqEb4rXgH53TICfZ:b-nejuin) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/security/auth.go#L101)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I like this design but I'd rename the method to something like `applyRootOrNodeDNFlag` to emphasize that it's about combining the command line flags with a maybe-existing role option (it's good to cover that case even though I think it's not required at this point).
</blockquote></details>

changed to `applyRootOrNodeDNFlag`
___
*[`pkg/security/auth.go` line 142 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4hl4k7jPGNDnBfqPN:-Nu4suSQ50J2WLj0SS1F:b374syp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/security/auth.go#L142)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Nit: Move the node subject functions up to be next to the root ones.
</blockquote></details>

moved up.


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell)

___
*[`pkg/cli/interactive_tests/test_distinguished_name_validation.tcl` line 43 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4mB5s5xpb3JULQtlr:-Nu57Btv2HgpKUWBlVnR:buc3097) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/cli/interactive_tests/test_distinguished_name_validation.tcl#L43)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

updated
</blockquote></details>

use of SUBJECT role option requires an enterprise license.


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
i think we could add a hint to this error message: https://github.com/cockroachdb/cockroach/blob/f822bc6557aadf7103973acfba5885d4814a6815/pkg/sql/roleoption/role_option.go#L224

something like:
```
					return errors.WithHintf(
						pgerror.Newf(pgcode.InvalidParameterValue, "role %q cannot have a SUBJECT", u),
						"use the --%s CLI flag to configure root",
						cliflags.RootCertDistinguishedName.Name,
					)
```


and one high-level question: was a design that uses cluster settings instead of CLI flags considered? i can see that it would be difficult to plumb the `Settings` object where it's needed, but cluster settings also have an `OnChange` callback that could do similar logic of setting a global `rootDN/nodeDN` variable. the advantage is that this can be configured without a node restart. see: https://github.com/cockroachdb/cockroach/blob/13719ea685e0ff0308ae20239b5ed20619b6898a/pkg/settings/common.go#L149

i think the big flaw with my suggestion is if the setting is needed during cluster/node initialization, then only setting this with a cluster setting is already "too late"

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-:-Nu5L5UG5-l8utJgoMKX:bjphg7l)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @souravcrl)

___
*[`pkg/security/auth.go` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5O4x1EZOGOTDb2DGK:-Nu5O4x2-sadAMYT9mDb:bk94yby) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L36)):*
> ```Go
> 
> var rootSubject struct {
> 	syncutil.RWMutex
> ```

nit: we have a useful convention to use the word `mu` to indicate that a field is meant to be accessed with a mutex. so this could look like:

```
var rootSubject struct {
  mu struct {
    syncutil.Mutex
    results []benchmarkResult
  }
}
```

if you think that's too verbose, then changing the name could still help (we have this convention in other parts of the code too):
```
var rootSubjectMu struct {
	syncutil.RWMutex
	dn *ldap.DN
}

```
___
*[`pkg/security/auth.go` line 78 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5Ofa-BwyS2GtXNnWg:-Nu5Ofa-BwyS2GtXNnWh:b-clkueg) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L78)):*
> ```Go
> func SetRootSubject(rootDNString string) error {
> 	if len(rootDNString) == 0 {
> 		return nil
> ```

should this either return an error, or call `UnsetRootSubject`? making it a no-op seems unintuitive
___
*[`pkg/security/auth.go` line 84 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5OphHC1pzHghH_4GE:-Nu5OphHC1pzHghH_4GF:b-rlpn4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L84)):*
> ```Go
> 		return errors.Errorf("invalid root cert distinguished name: %q", rootDNString)
> 	}
> 	rootSubject.Lock()
> ```

super nit: it's a good practice to always use `defer` to unlock:
```
	rootSubject.Lock()
	defer rootSubject.Unlock()
	rootSubject.dn = rootDN
```

or if you want to more exact control over when the lock is held, use an anonymous function:
```
func () {
  rootSubject.Lock()
  defer rootSubject.Unlock()
  rootSubject.dn = rootDN
}()
```
___
*[`pkg/security/auth.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5PJcO1xyNcLPaQyCr:-Nu5PJcO1xyNcLPaQyCs:bj12ba8) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L91)):*
> ```Go
> 
> func UnsetRootSubject() {
> 	rootSubject.Lock()
> ```

similar comment about `defer`
___
*[`pkg/security/auth.go` line 98 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5OmWQ95FOIGXdlJz2:-Nu5OmWQ95FOIGXdlJz3:b871dq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L98)):*
> ```Go
> func SetNodeSubject(nodeDNString string) error {
> 	if len(nodeDNString) == 0 {
> 		return nil
> ```

same as above comment about returning an error or calling Unset
___
*[`pkg/security/auth.go` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5POVM4sbvwSwI-GsX:-Nu5POVM4sbvwSwI-GsY:bj12ba8) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L106)):*
> ```Go
> 	nodeSubject.Lock()
> 	nodeSubject.dn = nodeDN
> 	nodeSubject.Unlock()
> ```

similar comment about `defer`
___
*[`pkg/security/auth.go` line 113 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5PP288ttOQrsmWi87:-Nu5PP288ttOQrsmWi88:bj12ba8) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L113)):*
> ```Go
> 	nodeSubject.Lock()
> 	nodeSubject.dn = nil
> 	nodeSubject.Unlock()
> ```

similar comment about `defer`
___
*[`pkg/security/auth.go` line 133 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5Per01vJQ9jpUj1lM:-Nu5Per15CwHh-GOVQPs:bj12ba8) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L133)):*
> ```Go
> 		nodeSubject.RLock()
> 		dn = nodeSubject.dn
> 		nodeSubject.RUnlock()
> ```

similar comment about `defer`
___
*[`pkg/security/auth.go` line 138 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5Pxpa61MNdelj3ZQF:-Nu5Pxpa61MNdelj3ZQG:b4ggcxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L138)):*
> ```Go
> }
> 
> func CheckCertDNMatchesRootDNorNodeDN(
> ```

nit: add a comment to explain this function and when to use it. (this is more important for exported functions, but is generally always a good idea)
___
*[`pkg/security/auth.go` line 147 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5PnT-4a8jSjpls6O3:-Nu5PnT-4a8jSjpls6O4:bj12ba8) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L147)):*
> ```Go
> 	nodeSubject.RLock()
> 	nodeDN := nodeSubject.dn
> 	nodeSubject.RUnlock()
> ```

similar comment about `defer`
___
*[`pkg/security/auth.go` line 154 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5QDlUEyeqvPQo7ADS:-Nu5QDlUEyeqvPQo7ADT:b-tlj6sr) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L154)):*
> ```Go
> 			return true, false
> 		}
> 		if rootDN != nil && certDN.Equal(rootDN) || nodeDN != nil && certDN.Equal(nodeDN) {
> ```

nit: add parentheses to make this check more clear:
```
	if (rootDN != nil && certDN.Equal(rootDN)) || (nodeDN != nil && certDN.Equal(nodeDN)) {
```


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> Reviewed 1 of 2 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @souravcrl)

___
*[`pkg/cli/interactive_tests/test_distinguished_name_validation.tcl` line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4llrE7YLNMjv8HYdV:-Nu7kZqK4PyoxEM8QYbw:b1c5x6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/cli/interactive_tests/test_distinguished_name_validation.tcl#L17)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

There was a case when the directory was not created and test proceeded with creating a ca certificate. So have added send and was waiting for `prompt` here.
</blockquote></details>

Hmm, OK. it seems like there should be some way to abort if a `system` command fails but I don't know how to do it. 
___
*[`pkg/security/auth.go` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5O4x1EZOGOTDb2DGK:-Nu7kjmY2knvQznz8z6q:b8nl3vi) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L36)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: we have a useful convention to use the word `mu` to indicate that a field is meant to be accessed with a mutex. so this could look like:

```
var rootSubject struct {
  mu struct {
    syncutil.Mutex
    results []benchmarkResult
  }
}
```

if you think that's too verbose, then changing the name could still help (we have this convention in other parts of the code too):
```
var rootSubjectMu struct {
	syncutil.RWMutex
	dn *ldap.DN
}

```
</blockquote></details>

We use this convention for larger structs where only some of the fields are protected by the mutex, but for a simple container for a single object I think this is fine as it is. The cert principal map in this file uses the same pattern. 
___
*[`pkg/security/auth.go` line 78 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5Ofa-BwyS2GtXNnWg:-Nu7l5Wo0zQTnF135I0_:bdoybn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L78)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should this either return an error, or call `UnsetRootSubject`? making it a no-op seems unintuitive
</blockquote></details>

Yes, that seems like a good idea. 


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @souravcrl)

___
*[`pkg/security/auth.go` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5O4x1EZOGOTDb2DGK:-Nu9x-UN2V6ZI5BTP68D:b-lalnbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L36)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We use this convention for larger structs where only some of the fields are protected by the mutex, but for a simple container for a single object I think this is fine as it is. The cert principal map in this file uses the same pattern.
</blockquote></details>

adding a `Mu` suffix to the name still seems like a very unobtrusive, but possibly helpful hint to future engineers working in this code, but this is just a suggestion, not a blocking comment


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @souravcrl)

___
*[`pkg/security/auth.go` line 155 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu9y0h72c6i4T1VzUoD:-Nu9y0h72c6i4T1VzUoE:bwdvpt4) ([raw file](https://github.com/cockroachdb/cockroach/blob/189fe8cb379ba7d87e14a7f56b96eab44ca70158/pkg/security/auth.go#L155)):*
> ```Go
> 			return rootOrNodeDNSet, certDNMatchesRootOrNodeDN
> 		}
> 		if (rootDN == nil || certDN.Equal(rootDN)) && (nodeDN == nil || certDN.Equal(nodeDN)) {
> ```

the `||` and `&&` seems mixed up now


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
We will still need a cluster setting to mandatorily enforce if subjectRoleOption and rootDN, nodeDN are set (`server.client.cert.subject.required` ) which will be needed during cluster/node initialization but if we are to use cluster setting for `rootDN/nodeDN` then I think we might risk these to be misconfigured where customer doesn't actually have the certs with required subjects and then there is no way to start/access the cluster. 

PTAL: I have updated implementation for rootDN, nodeDN and corresponding methods.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-:-NuDElqw-4cOt7RbmKd8:b-bfcuo6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @rafiss, and @souravcrl)

___
*[`pkg/cli/interactive_tests/test_distinguished_name_validation.tcl` line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4llrE7YLNMjv8HYdV:-NuJQ5ttAjqd_I2HJF4M:b52ulxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/cli/interactive_tests/test_distinguished_name_validation.tcl#L17)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Hmm, OK. it seems like there should be some way to abort if a `system` command fails but I don't know how to do it.
</blockquote></details>

we can use the exit code by using catch but `system` command does not return executed commands standard output. ref: http://osr507doc.xinuos.com/en/man/html.TCL/system.TCL.html
something like this: 
```
set directory "my-directory"
set command "mkdir $directory"

# Execute the system command
set exit_code [catch {system $command} output]

if {$exit_code == 0} {
    puts "Directory '$directory' created successfully"
} else {
    puts "Failed to create directory '$directory'"
}
``` 
but for using `expect` or `eexpect` which can wait for certain output for a fixed time period, we have to `spawn` a new process or do `send` to bash process created at beginning of the program. I have done the latter here.
___
*[`pkg/security/auth.go` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5O4x1EZOGOTDb2DGK:-NuKj9E04l0kvQglBtru:b-gczymz) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L36)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

adding a `Mu` suffix to the name still seems like a very unobtrusive, but possibly helpful hint to future engineers working in this code, but this is just a suggestion, not a blocking comment
</blockquote></details>

added a struct `userCertDistinguishedNameMu` for rootDN and nodeDN struct variables.
___
*[`pkg/security/auth.go` line 78 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5Ofa-BwyS2GtXNnWg:-NuKj_Wx4pGJhoyYsMpx:b-y3q0mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L78)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yes, that seems like a good idea.
</blockquote></details>

updated to call `UnsetRootSubject`.
___
*[`pkg/security/auth.go` line 84 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5OphHC1pzHghH_4GE:-NuKjimdAc5xSB_ZlWqe:bcrqivy) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L84)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: it's a good practice to always use `defer` to unlock:
```
	rootSubject.Lock()
	defer rootSubject.Unlock()
	rootSubject.dn = rootDN
```

or if you want to more exact control over when the lock is held, use an anonymous function:
```
func () {
  rootSubject.Lock()
  defer rootSubject.Unlock()
  rootSubject.dn = rootDN
}()
```
</blockquote></details>

added method setDN to struct type `userCertDistinguishedNameMu` which handles this.
___
*[`pkg/security/auth.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5PJcO1xyNcLPaQyCr:-NuKk4vS4CZZwNzIn9p5:bpibq0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L91)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

similar comment about `defer`
</blockquote></details>

added method unsetDN to struct type `userCertDistinguishedNameMu` which handles this.
___
*[`pkg/security/auth.go` line 98 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5OmWQ95FOIGXdlJz2:-NuKkDp8BFeaq5oaynrV:bq88eis) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L98)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

same as above comment about returning an error or calling Unset
</blockquote></details>

added a common method setDNWithString which addresses this.
___
*[`pkg/security/auth.go` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5POVM4sbvwSwI-GsX:-NuKkKz2-NoY0BpjzNnE:b-1qp9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L106)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

similar comment about `defer`
</blockquote></details>

added a method getDN for struct which addresses this
___
*[`pkg/security/auth.go` line 113 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5PP288ttOQrsmWi87:-NuKkR7nAyELcW-HnLN4:b-wdz0d1) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L113)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

similar comment about `defer`
</blockquote></details>

added a method unsetDN for struct which addresses this
___
*[`pkg/security/auth.go` line 133 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5Per01vJQ9jpUj1lM:-NuKkTFOAGO8fTT3a3Kz:b-5mugkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L133)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

similar comment about `defer`
</blockquote></details>

added a method setDN for struct which addresses this
___
*[`pkg/security/auth.go` line 138 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5Pxpa61MNdelj3ZQF:-NuKt2tGCliaORVw5L8W:b-3loa35) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L138)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: add a comment to explain this function and when to use it. (this is more important for exported functions, but is generally always a good idea)
</blockquote></details>

added comment to explain the fn.
___
*[`pkg/security/auth.go` line 147 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5PnT-4a8jSjpls6O3:-NuKkcVECJQyfnCo4GG-:b-1qp9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L147)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

similar comment about `defer`
</blockquote></details>

added a method getDN for struct which addresses this
___
*[`pkg/security/auth.go` line 154 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5QDlUEyeqvPQo7ADS:-NuKkfoaDh_lGky3woFq:bctx901) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L154)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: add parentheses to make this check more clear:
```
	if (rootDN != nil && certDN.Equal(rootDN)) || (nodeDN != nil && certDN.Equal(nodeDN)) {
```
</blockquote></details>

added parathesis
___
*[`pkg/security/auth.go` line 155 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu9y0h72c6i4T1VzUoD:-NuKkn6PBafQeLsG8xfO:b-e35328) ([raw file](https://github.com/cockroachdb/cockroach/blob/189fe8cb379ba7d87e14a7f56b96eab44ca70158/pkg/security/auth.go#L155)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the `||` and `&&` seems mixed up now
</blockquote></details>

updated the implementation and added a comment to explain this.


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
Now I am able to pass all CI tests except Bazel Essential because it does not have the required configurations. There is an ongoing PR to remove interactive(acceptance) tests from Bazel Essential here: https://github.com/cockroachdb/cockroach/pull/121481

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-:-NubpUPR1Hjkl61TYnEC:b-jeafs6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @rafiss, and @souravcrl)



<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> was a design that uses cluster settings instead of CLI flags considered?

Cluster settings won't work for the `node` user because a new node cannot learn anything about cluster settings until it has successfully mutually authenticated to its peers as `node`. Additionally, there's the risk Sourav noted that misconfigured cluster settings can lock you out of the cluster entirely; with root auth controlled by a command line flag you can always restart the node to make your certs work. 

> Reviewed 7 of 9 files at r6, 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-:-Nugm47nABTO4hQB600I:bme9gbf)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @souravcrl)

___
*[`pkg/cli/interactive_tests/test_distinguished_name_validation.tcl` line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu4llrE7YLNMjv8HYdV:-NugoJXCCQ0AVr5PJhP2:b-u2ssqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4aa11eb11894524b073cdea559f52137525e4e4/pkg/cli/interactive_tests/test_distinguished_name_validation.tcl#L17)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

we can use the exit code by using catch but `system` command does not return executed commands standard output. ref: http://osr507doc.xinuos.com/en/man/html.TCL/system.TCL.html
something like this: 
```
set directory "my-directory"
set command "mkdir $directory"

# Execute the system command
set exit_code [catch {system $command} output]

if {$exit_code == 0} {
    puts "Directory '$directory' created successfully"
} else {
    puts "Failed to create directory '$directory'"
}
``` 
but for using `expect` or `eexpect` which can wait for certain output for a fixed time period, we have to `spawn` a new process or do `send` to bash process created at beginning of the program. I have done the latter here.
</blockquote></details>

Hmm, in the OCSP test I used spawn/wait when i expected a non-zero exit status but assumed that system would fail on its own if the process failed. If system is ignoring status codes then the ocsp test is covering a lot less than I thought it was. 
___
*[`pkg/security/auth.go` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5O4x1EZOGOTDb2DGK:-Nugp1YHEzinp2Q4mMdh:b-s14v87) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L36)):*
<details><summary><i>Previously, souravcrl wrote…</i></summary><blockquote>

added a struct `userCertDistinguishedNameMu` for rootDN and nodeDN struct variables.
</blockquote></details>

The point of the `Mu` suffix (or the `mu` inner struct) is that the `mu` is visible at the point of reference so you can't forget to lock/unlock it. Putting `Mu` on the type while the variable name is `nodeSubject` and `rootSubject` doesn't accomplish this. Anyway, I don't want to pick this code apart any further; it's fine now. 


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
once the tests are green, this lgtm! i'll be out for most of this week, so don't want to block this

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-:-NuqysvL3ZlqspYfabO0:b-f12s3y)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @souravcrl)

___
*[`pkg/security/auth.go` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5O4x1EZOGOTDb2DGK:-Nuqxh8E1DRHBFXMFqiI:b-6p2gb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L36)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The point of the `Mu` suffix (or the `mu` inner struct) is that the `mu` is visible at the point of reference so you can't forget to lock/unlock it. Putting `Mu` on the type while the variable name is `nodeSubject` and `rootSubject` doesn't accomplish this. Anyway, I don't want to pick this code apart any further; it's fine now.
</blockquote></details>

i see, my mistake. i agree it only makes sense on a variable name, and i'd confused myself when reading this


<!-- Sent from Reviewable.io -->

souravcrl reviewed (commented):
The changes to CI pipeline got merged and my CI tests are passing. There is a flaky test in experimental roach test but not related to my change. The thread related to secrets part that was updated in the PR: https://cockroachlabs.slack.com/archives/CJ0H8Q97C/p1712345082826039?thread_ts=1711979995.558999&cid=CJ0H8Q97C

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-:-NuwnZSz-Sxu9z6qh4SJ:b9e3svv)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bdarnell and @rafiss)

___
*[`pkg/security/auth.go` line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120786#-Nu5O4x1EZOGOTDb2DGK:-NuwlDTC0ejhnBCnJEJ1:b-x1ffjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f39942b6045e6839b87ff4709eca26efad9b90/pkg/security/auth.go#L36)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i see, my mistake. i agree it only makes sense on a variable name, and i'd confused myself when reading this
</blockquote></details>

I have updated variable names to have `Mu` suffix.


<!-- Sent from Reviewable.io -->

rafiss commented:
bors r+
